dscribe/descriptors/mbtr.py

```diff
@@ -1104,3 +1105,467 @@ def _get_k3(self, system):
+            return (n_indices, 3, n_features), (n_features,)
+
+        derivatives_shape, descriptor_shape = get_shapes(inp[0])
+
+        def is_variable(inp):
+            for job in inp[1:]:
+                i_der_shape, i_desc_shape = get_shapes(job)
+                if i_der_shape != derivatives_shape or i_desc_shape != descriptor_shape:
+                    return True
+            return False
+
+        if is_variable(inp):
+            derivatives_shape = None
+            descriptor_shape = None
+
+        # Create in parallel
+        output = self.derivatives_parallel(
+            inp,
+            self.derivatives_single,
+            n_jobs,
+            derivatives_shape,
+            descriptor_shape,
+            return_descriptor,
+            only_physical_cores,
+            verbose=verbose,
+        )
+
+        return output
+
+    # Modeled after create_single() function but uses _get_k1_derivatives(),
+    # _get_k2_derivatives(), and _get_k3_derivatives() functions instead of
+    # _get_k1(), _get_k2(), and _get_k3() to also calculate the derivatives.
+    def derivatives_single(
+        self,
+        system,
+        indices,
+        method="analytical",
+        return_descriptor=True,
+    ):
+        """Return the derivatives for the given system.
+
+        Args:
+            system (:class:`ase.Atoms`): Atomic structure.
+            indices (list): Indices of atoms for which the derivatives will be
+                computed for.
+            method (str): The method for calculating the derivatives. Supports
+                'numerical'.
+            return_descriptor (bool): Whether to also calculate the descriptor
+                in the same function call. This is true by default as it
+                typically is faster to calculate both in one go.
+
+        Returns:
+            If return_descriptor is True, returns a tuple, where the first item
+            is the derivative array and the second is the descriptor array.
+            Otherwise only returns the derivatives array. The derivatives array
+            is a 3D numpy array. The dimensions are: [n_atoms, 3, n_features].
+            The first dimension goes over all the atoms in the system. The
+            second dimension goes over the cartesian components, x, y and z.
+            The last dimension goes over the features in the default order.
+        """
+
+        # Check that the derivative calculations are supported for the used
+        # MBTR parameters
+        supported_normalization = ["none", "n_atoms"]
+        if self.normalization not in supported_normalization:
+            raise ValueError(
+                "Derivatives not implemented for normalization option '{}'. Please choose from: {}".format(
+                    self.normalization, supported_normalization
+                )
+            )
+
+        if self.flatten == False:
+            raise ValueError("Derivatives not implemented for flatten=False.")
```

Comment:
Moved them. Now the checks don't apply to `derivatives_single`-function. I guess that this is not a problem because `derivatives_single` is not supposed to be used from outside the class? If so, should this be indicated somehow to the user (e.g. renaming the function to `_derivatives_single`)?

On a related note. Most of the functionality in `derivatives`-function is shared between all global descriptors. In my opinion a more intuitive code structure would implement `derivatives` in a base class. Then, every descriptor would only need to implement `derivatives_single`-function, and the descriptor specific error checks would also be there.

Reply:
Yes, you're absolutely right. It is currently quite a mess to understand which of the classes - base or subclass - needs to implement the functions. I already started changing this a bit in v1.2.3 and will take care of migrating your code.